source/extensions/access_loggers/grpc/grpc_access_log_impl.h

```diff
@@ -22,153 +25,128 @@ namespace GrpcCommon {
+    InflightMessageTtlManager(Event::Dispatcher& dispatcher,
+                              CriticalAccessLoggerGrpcClientStats& stats,
+                              Grpc::BufferedAsyncClient<RequestType, ResponseType>& client,
+                              std::chrono::milliseconds message_ack_timeout)
+        : dispatcher_(dispatcher), message_ack_timeout_(message_ack_timeout) {
+      timer_ = dispatcher_.createTimer([this, &client, &stats] {
+        const auto now = dispatcher_.timeSource().monotonicTime();
+        std::vector<MonotonicTime> expired_timepoints;
+        std::set<uint32_t> expired_message_ids;
+
+        auto it = deadline_.lower_bound(now);
+        while (it != deadline_.end()) {
+          for (auto&& id : it->second) {
+            expired_message_ids.emplace(id);
+          }
+          expired_timepoints.push_back(it->first);
+          ++it;
+        }
 
-    if (active_stream_->stream_ == nullptr ||
-        active_stream_->stream_.isAboveWriteBufferHighWatermark()) {
-      active_stream_.reset();
-      return;
-    }
+        for (auto&& id : expired_message_ids) {
+          const auto& message_buffer = client.messageBuffer();
 
-    uint32_t id = MessageUtil::hash(message);
-    const auto message_byte_size = message.ByteSizeLong();
+          if (message_buffer.find(id) == message_buffer.end()) {
+            continue;
+          }
 
-    if (current_critical_buffer_size_bytes_ + message_byte_size > max_critical_buffer_size_bytes_) {
-      stats_.ciritcal_logs_disposed_.inc();
-      return;
-    }
+          auto& message = message_buffer.at(id);
+          if (message.first == Grpc::BufferState::Pending) {
+            client.onError(id, true);
+            stats.critical_logs_message_timeout_.inc();
+          }
+        }
 
-    buffered_messages_[id] = BufferedMessage{nullptr, BufferedMessage::State::Initial, message};
-    current_critical_buffer_size_bytes_ += message_byte_size;
-    stats_.pending_critical_logs_.inc();
-
-    for (auto&& buffered_message : buffered_messages_) {
-      uint32_t id = buffered_message.first;
-      buffered_message.second.message_.set_id(id);
-      buffered_message.second.state_ = BufferedMessage::State::Pending;
-
-      buffered_message.second.timer_ = dispatcher_.createTimer([&]() {
-        if (buffered_message.second.state_ == BufferedMessage::State::Pending) {
-          stats_.pending_timeout_.inc();
-          buffered_message.second.state_ = BufferedMessage::State::Initial;
-          buffered_message.second.timer_->disableTimer();
-          buffered_message.second.timer_.reset();
+        for (auto&& timepoint : expired_timepoints) {
+          deadline_.erase(timepoint);
         }
+
+        timer_->enableTimer(message_ack_timeout_);
       });
-      buffered_message.second.timer_->enableTimer(message_ack_timeout_);
 
-      active_stream_->stream_->sendMessage(buffered_message.second.message_, false);
+      timer_->enableTimer(message_ack_timeout_);
```

Comment:
I think this would result in the timer firing periodically even when there are no pending messages, maybe we should gate this on there being any messages?

I also think that since we just evaluate the timeouts every N seconds we might not accurately enforce timeouts? Don't we need to set the time to the time between now and the next deadline to expire?

Reply:
I think it is similar to ttl.h's implementation because it restarts the timer at the interval between the earliest time point of the deadline and the current time. By the way, I checked the ttl.h implementation, and I don't think it takes this corner case into account, but can you guarantee that such a case will not occur?

https://github.com/envoyproxy/envoy/pull/17486/files#diff-b0bbeea52042998049ac1211b5c7203366ea23f112c0d4f4d939f04697bd0448R121-R 125

Reply:
Fixed. I think it is similar to ttl.h's implementation because it restarts the timer at the interval between the earliest time point of the deadline and the current time.